### PR TITLE
Implement MSC3819: Allowing widgets to send/receive to-device messages

### PR DESCRIPTION
**Pull request by robintown**
https://github.com/matrix-org/matrix-spec-proposals/pull/3819

### CODE REVIEW

**Comment by turt2live** [comment]
> /me doesn't feel like he knows enough to review this one 😅

it's a great opportunity to learn then :) that's why the review requests are randomly assigned.

**Comment by SimonBrandner** [comment]
> > /me doesn't feel like he knows enough to review this one sweat_smile
> 
> it's a great opportunity to learn then :) that's why the review requests are randomly assigned.

felt like stalling the review :sweat_smile: though yes, there are multiple people assigned here

File: src/models/WidgetEventCapability.ts [hunk 1]
```diff
@@ -25,14 +31,15 @@ export class WidgetEventCapability {
...
-        public readonly isState: boolean,
+        public readonly kind: EventKind,
         public readonly keyStr: string | null,
         public readonly raw: string,
     ) {
     }
 
-    public matchesAsStateEvent(eventType: string, stateKey: string): boolean {
-        if (!this.isState) return false; // looking for state, not state
+    public matchesAsStateEvent(direction: EventDirection, eventType: string, stateKey: string): boolean {
```
Comment (SimonBrandner) [inline src/models/WidgetEventCapability.ts:40]: I am assuming this is a breaking change?

File: src/ClientWidgetApi.ts [hunk 2]
```diff
@@ -453,6 +462,32 @@ export class ClientWidgetApi extends EventEmitter {
         });
     }
 
+    private async handleSendToDevice(request: ISendToDeviceFromWidgetActionRequest): Promise<void> {
```
Comment (SimonBrandner) [inline src/ClientWidgetApi.ts:465]: My personal preference is to get rid of the `else` and add `return`s to the if statements, but it probably is just that - a personal preference

**Review by SimonBrandner — APPROVED** [review]
Looks like it does what it should do; not that I would feel confident about understanding it

File: src/models/WidgetEventCapability.ts (hunk 1 above)
Comment (robintown) [reply src/models/WidgetEventCapability.ts:40]: Yes, it is. I've asked Travis, and it seems some light breakages like this are okay given the current maturity of the package

File: src/ClientWidgetApi.ts (hunk 2 above)
Comment (robintown) [reply src/ClientWidgetApi.ts:465]: Well, if it's all the same to you, I'm going to keep it this way :stuck_out_tongue:

**Pull request merged**